### ISSUE TITLE
game crash: attack_prediction assertion fail during AI turn

### ISSUE THREAD

**Issue by vgaming** [wesnoth/wesnoth#5002]
Game crashes during AI turn with:
```log
wesnoth: src/attack_prediction.cpp:1668: static void {anonymous}::monte_carlo_combat_matrix::scale_probabilities(const std::vector<double>&, std::vector<double>&, double, unsigned int): Assertion `std::abs(std::accumulate(target.begin(), target.end(), 0.0) - 1.0) < 0.001' failed.
```

Possibly related fact: there were a lot of units that had a lot of strikes/damage before the AI turn.

The game is run on ArchLinux, `v1.15.3+dev (4becc858f12-Clean)`

Any additional information that could help investigating that? I remember reporting a similar bug some time ago BTW - maybe we forgot to port the change to the `master` branch?

**Comment by vgaming** [comment]
@jyrkive I know you love it when I'm raising `attack_prediction.cpp` assertion bugs :D

**Comment by stevecotton** [comment]
@vgaming do you have an autosave that can reliably reproduce the bug?

**Comment by vgaming** [comment]
@stevecotton nope, unfortunately no. I tried to reload it a few times, but didn't hit the bug again..
[Creep Wars 3team 1.5.2 Turn 71-d.gz](https://github.com/wesnoth/wesnoth/files/4967879/Creep.Wars.3team.1.5.2.Turn.71-d.gz)

**Comment by jyrkive** [comment]
Hi.

Sorry it took me so long to start investigating this. My job has been draining all my energy and there hasn't been any left over for Wesnoth.

Unfortunately the bug report isn't actionable without a way to reproduce the bug. I tried simply pressing "end turn" with the save game you attached (repeated ten times) and didn't get the crash. The assertion only detects the situation where the sum of hit probabilities doesn't add up to 100%, but it doesn't give me any clues for the underlying cause. I'm afraid there is nothing I can do. :(

**Comment by vgaming** [comment]
@jyrkive Hey, is there any way to force AI be deterministic/reproducible in the future? Maybe some flag that forces the usage of a different RNG? The results might still look "random" if the RNG is seeded e.g. play field information each turn, or wesnoth Lua/WML random state each turn, or a random (non-deterministic) value for each game that is then written down e.g. as a wesnoth variable and re-used later.

Having reproducibility would help here, because identical savegames would mean identical AI decisions, and ultimately identical crashes.

To attack the problem from another angle, do you see any more details that could be logged while crashing here? Is there any "object" that wesnoth could log fully and that would represent the _full_ information about the attacker and the defender?

**Comment by vgaming** [comment]
@jyrkive e.g. if one of the "callers" is a pure function with only 2 arguments, is there a possibility to log both the arguments fully?

**Comment by jyrkive** [comment]
I think the main reason why the AI is nondeterministic is that its behavior depends on outcomes of earlier battles. In other words, the AI makes some decisions, fights some battles, then makes more decisions. Now *the game state at the time of the latter decisions depends on how the earlier battles went!*

That issue probably doesn't exist if the option to save the random seed in save files is used.

> To attack the problem from another angle, do you see any more details that could be logged while crashing here? Is there any "object" that wesnoth could log fully and that would represent the full information about the attacker and the defender?

Yeah, there is additional information that could be logged to help investigation. Most notably *by how much* the variation from 100% exceeds the allowed threshold (0,1 %). Some variation is allowed simply because floating point math isn't fully accurate and thus the probabilities will almost never add up to exactly 100%.

If the threshold is only barely exceeded, the right fix is just making it larger - the explanation is just that the math is less accurate than I thought. If, instead, the sum is completely wrong (like 12,5 % in the case I fixed in commit 77da93cbf2a5686f4107b732780069de6506c2e7), it means that calculations in some corner case are just wrong.

**Comment by gfgtdf** [comment]
> That issue probably doesn't exist if the option to save the random seed in save files is used.

The 'save random seed' only applies to the 'synced' rng, and explicitly not to ai decision making.

**Comment by vgaming** [comment]
> That issue probably doesn't exist if the option to save the random seed in save files is used.

Do I understand it correctly that I can specify something for wesnoth that the same game events would produce the same battle outcomes by default? Where can I find this option, I'd like to enable it...

Unrelated, once this option is set, would the AI be able to pick it up and also make itself repro?

> Most notably by how much the variation from 100% exceeds

I'm not sure if my proposal got across here. I mean to ask, is it possible to log the attacker-s damage, strikes, health, poison/slow/whatever -- everything -- to be able to reproduce the issue at code level? E.g. a game crash would then log ALL the details, and you could manually call the C++ function with those details for probability calculation as a developer and observe the crash.

**Comment by jyrkive** [comment]
> Do I understand it correctly that I can specify something for wesnoth that the same game events would produce the same battle outcomes by default? Where can I find this option, I'd like to enable it...

This option:

![image](https://user-images.githubusercontent.com/1665617/95059004-fa8ef780-0700-11eb-8176-b2ec2abd3fa3.png)

> I'm not sure if my proposal got across here. I mean to ask, is it possible to log the attacker-s damage, strikes, health, poison/slow/whatever -- everything -- to be able to reproduce the issue at code level? E.g. a game crash would then log ALL the details, and you could manually call the C++ function with those details for probability calculation as a developer and observe the crash.

Yes, logging that all would be possible, but such logs would be extremely verbose. Maybe they could be enabled only with a dedicated attack prediction log domain...

**Comment by vgaming** [comment]
Thanks for the "random seed" explanation BTW! Unfortunately only applies to Campaigns it seems..

**Comment by vgaming** [comment]
(Just in case, I've never actually programmed in C++ more than a few lines, I don't know how it works exactly. Writing from a perspective of different programming languages like Java/Scala or I think Lua as well)

**Comment by jyrkive** [comment]
> @jyrkive is it possible to catch the exception in the code that has all the variables in scope, add them as details to the stack and throw the exception further out?

Unfortunately the information has been lost at that point. To be able to reproduce such a crash, all simulated battles for the unit in question need to be repeated (three is the maximum number of battles the AI can simulate for one unit, and also almost always the actual number when attack prediction crashes). At the time of such a crash, only the info for the final battle is available.

**Comment by mattsc** [comment]
I just found this issue.  FYI, #5533 is another report of the same error.  I've been able to track it down in that case and will (hopefully) have it fixed soon.  There's a good chance (but no guarantee, of course) that the cause is the same, so I suggest closing this issue when #5533 is fixed, given that there is no way to reproduce what exactly caused the error in this report.

**Comment by vgaming** [comment]
@mattsc Hey, thanks for noticing this. I'm fine with closing this issue if you believe it's likely that the source of it will be closed along with #5533.
The situation in this issue was not reproducible for me anyway, as I write in one of the comments above, so closing seems like a good choice (after we fix #5533)

**Comment by mattsc** [comment]
#5533 is now fixed via a5b714e5bd
As discussed, I am closing this PR as well.
